zerver/lib/users.py

```diff
@@ -342,8 +342,59 @@ def access_user_by_email(
     allow_bots: bool = False,
     for_admin: bool,
 ) -> UserProfile:
+    """
+    Fetch a user by email address. Endpoints using this function can be queried either with:
+    1) The real email address of the intended user, if the requester believes that user exists and allows
+       their email address to be visible to the requester.
+    2) The dummy email address (of the approximate shape 'user<id>@<dummy email domain>') of the intended
+       user. We detect when the format of the provided email address matches the format of our dummy
+       email addresses and extract the user id from it for a regular id-based lookup.
+
+       In particular, this mode is kept around for backwards compatibility with the old behavior of the
+       `GET /users/{email}` endpoint, which required use of the dummy email address for lookups of any user
+       which didn't have EMAIL_ADDRESS_VISIBILITY_EVERYONE set, regardless of how the actual
+       email_address_visibility setting related to the role of the requester.
+
+       Note: If the realm.host value changes (e.g. due to the server moving to a new
+       domain), the required dummy email values passed here will need to be updated
+       accordingly to match the new value. This deviates from the original API behavior,
+       where the lookups were supposed to match the UserProfile.email value, which was
+       **not** updated for existing users even if the server moved domains. See
+       get_fake_email_domain for details of how the email domain for dummy email addresses
+       is determined.
```

Comment:
I feel that the `.email` behavior, where you can move the realm or server to a different domain/subdomain, and old users will stay with their `.email` set to dummy addresses in the **old** domain, while new users get created in the **new** domain is weird enough, that this small break of backward compatibility is justified.

with this new API approach, you just send dummy addresses formatted based on the current `realm.host`, which is more logical and easier than "you have to figure out what the `UserProfile.email` is"

Reply:
Yeah, I don't think it's particularly intentional that `.email` stays the same after domain changes... though it is required for bot users, because of how their API keys are structured.